crates/pallet-domains/src/lib.rs

```diff
@@ -1595,6 +1603,57 @@ impl<T: Config> Pallet<T> {
         Ok(())
     }
 
+    fn check_slot_and_proof_of_time(
+        slot_number: u64,
+        proof_of_time: PotOutput,
+        pre_dispatch: bool,
+    ) -> Result<(), BundleError> {
+        // NOTE: the `current_block_number` from `frame_system` is initialized during `validate_unsigned` thus
+        // it is the same value in both `validate_unsigned` and `pre_dispatch`
+        let current_block_number = frame_system::Pallet::<T>::current_block_number();
+
+        // Check if the slot is in future
+        //
+        // NOTE: during `validate_unsigned` this is implicitly checked within `is_proof_of_time_valid` since we
+        // are using quick verification which will return `false` if the `proof-of-time` is not seem by the node
+        // before.
+        if pre_dispatch {
+            if let Some(future_slot) = T::BlockSlot::future_slot(current_block_number) {
+                ensure!(slot_number <= *future_slot, BundleError::SlotInTheFuture)
+            }
+        }
+
+        // Check if the bundle is built too long time ago and beyond `T::BundleLongevity` number of consensus blocks.
+        let produced_after_block_number = T::BlockSlot::slot_produced_after(slot_number.into());
+        let produced_after_block_hash = if produced_after_block_number == current_block_number {
+            // The hash of the current block is only available in the next block thus use the parent hash here
+            frame_system::Pallet::<T>::parent_hash()
+        } else {
+            frame_system::Pallet::<T>::block_hash(produced_after_block_number)
+        };
```

Comment:
maybe there should be a const assertion to ensure the BlockSlotCount is always less than frame_system block hash count